packages/service-library/src/servicelib/deferred_tasks/_base_deferred_handler.py

```diff
@@ -0,0 +1,103 @@
+from abc import ABC, abstractmethod
+from datetime import timedelta
+from typing import Any, ClassVar, Generic, TypeAlias, TypeVar
+
+from pydantic import NonNegativeInt
+
+from ._models import ClassUniqueReference, TaskResultError, TaskUID
+
+ResultType = TypeVar("ResultType")
+
+StartContext: TypeAlias = dict[str, Any]
+GlobalsContext: TypeAlias = dict[str, Any]
+
+# composed by merging `GlobalsContext` and `StartContext`
+DeferredContext: TypeAlias = dict[str, Any]
+
+
+class BaseDeferredHandler(ABC, Generic[ResultType]):
+    """Base class to define a deferred task."""
+
+    SUBCLASSES: ClassVar[list[type["BaseDeferredHandler"]]] = []
+
+    def __init_subclass__(cls, **kwargs):
+        super().__init_subclass__(**kwargs)
+        BaseDeferredHandler.SUBCLASSES.append(cls)
+
+    @classmethod
+    def _get_class_unique_reference(cls) -> ClassUniqueReference:
+        """
+        returns: a unique reference for this class (module and name)
+        """
+        return f"{cls.__module__}.{cls.__name__}"
+
+    @classmethod
+    async def get_retries(cls, context: DeferredContext) -> NonNegativeInt:
+        """
+        returns: the amount of retries in case of error (default: 0)
+
+        This is used only when ``run_deferred`` raises an error other than `asyncio.CancelledError`` and this
+        value is > 0. The code inside ``run_deferred`` will be retried.
+
+        NOTE: if the process running the ``run_deferred`` code dies, it automatically gets
+        retried when the process is restarted or by another copy of the service.
+        """
+
+        _ = context
+        return 0
+
+    @classmethod
+    @abstractmethod
+    async def get_timeout(cls, context: DeferredContext) -> timedelta:
+        """return the timeout for the execution of `run_deferred`.
+        If ``run_deferred`` does not finish executing in time a timeout exception will be raised
+        """
+
+    @classmethod
+    @abstractmethod
+    async def start_deferred(cls, **kwargs) -> StartContext:
```

Comment:
MINOR: avoid `deferred` in name since it is self-? `start`/`on_created`, `run` etc